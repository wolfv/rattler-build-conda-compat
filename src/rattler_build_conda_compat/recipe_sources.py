from __future__ import annotations

import typing
from collections.abc import MutableMapping
from dataclasses import dataclass
from typing import Any, List, Union, cast

from rattler_build_conda_compat.jinja.jinja import (
    RecipeWithContext,
    jinja_env,
    load_recipe_context,
)
from rattler_build_conda_compat.loader import _eval_selector
from rattler_build_conda_compat.variant_config import variant_combinations

from .conditional_list import ConditionalList, visit_conditional_list

if typing.TYPE_CHECKING:
    from collections.abc import Iterator


OptionalUrlList = Union[str, List[str], None]


@dataclass(frozen=True)
class Source:
    url: str | list[str]
    template: str | list[str]
    context: dict[str, str] | None = None
    sha256: str | None = None
    md5: str | None = None

    def __getitem__(self, key: str) -> str | list[str] | None:
        return self.__dict__[key]

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, Source):
            return NotImplemented
        return (self.url, self.sha256, self.md5) == (other.url, other.sha256, other.md5)

    def __hash__(self) -> int:
        return hash((tuple(self.url), self.sha256, self.md5))


def get_all_sources(recipe: MutableMapping[str, Any]) -> Iterator[MutableMapping[str, Any]]:
    """
    Get all sources from the recipe. This can be from a list of sources,
    a single source, or conditional and its branches.

    Arguments
    ---------
    * `recipe` - The recipe to inspect. This should be a yaml object.

    Returns
    -------
    A list of source objects.
    """
    sources = recipe.get("source", None)
    sources = typing.cast(ConditionalList[MutableMapping[str, Any]], sources)

    # Try getting all url top-level sources
    if sources is not None:
        source_list = visit_conditional_list(sources, None)
        for source in source_list:
            yield source

    cache_output = recipe.get("cache", None)
    if cache_output is not None:
        sources = cache_output.get("source", None)
        sources = typing.cast(ConditionalList[MutableMapping[str, Any]], sources)
        if sources is not None:
            source_list = visit_conditional_list(sources, None)
            for source in source_list:
                yield source

    outputs = recipe.get("outputs", None)
    if outputs is None:
        return

    outputs = visit_conditional_list(outputs, None)
    for output in outputs:
        sources = output.get("source", None)
        sources = typing.cast(ConditionalList[MutableMapping[str, Any]], sources)
        if sources is None:
            continue
        source_list = visit_conditional_list(sources, None)
        for source in source_list:
            yield source


def get_all_url_sources(recipe: MutableMapping[str, Any]) -> Iterator[str]:
    """
    Get all url sources from the recipe. This can be from a list of sources,
    a single source, or conditional and its branches.

    Arguments
    ---------
    * `recipe` - The recipe to inspect. This should be a yaml object.

    Returns
    -------
    A list of URLs.
    """

    def get_first_url(source: MutableMapping[str, Any]) -> str:
        if isinstance(source["url"], list):
            return source["url"][0]
        return source["url"]

    return (get_first_url(source) for source in get_all_sources(recipe) if "url" in source)


def render_all_sources(  # noqa: C901
    recipe: RecipeWithContext,
    variants: list[dict[str, list[str]]],
    override_version: str | None = None,
) -> set[Source]:
    """
    This function should render _all_ URL sources from the given recipe and with the given variants.
    Variants can be loaded with the `variant_config.variant_combinations` module.
    Optionally, you can override the version in the recipe context to render URLs with a different version.
    """

    def render(template: str | list[str], context: dict[str, str]) -> str | list[str]:
        if isinstance(template, list):
            return [cast(str, render(t, context)) for t in template]
        template = env.from_string(template)
        return template.render(context_variables)

    if override_version is not None:
        recipe["context"]["version"] = override_version

    final_sources = set()
    for v in variants:
        combinations = variant_combinations(v)
        for combination in combinations:
            env = jinja_env(combination)

            context = recipe.get("context", {})
            # render out the context section and retrieve dictionary
            context_variables = load_recipe_context(context, env)

            # now evaluate the if / else statements
            sources = recipe.get("source")
            if sources:
                if not isinstance(sources, list):
                    sources = [sources]

                for elem in visit_conditional_list(
                    sources,
                    lambda x, combination=combination: _eval_selector(x, combination),  # type: ignore[misc]
                ):
                    # we need to explicitly cast here
                    elem_dict = typing.cast(dict[str, Any], elem)
                    sha256, md5 = None, None
                    if elem_dict.get("sha256") is not None:
                        sha256 = typing.cast(str, render(str(elem_dict["sha256"]), context_variables))
                    if elem_dict.get("md5") is not None:
                        md5 = typing.cast(str, render(str(elem_dict["md5"]), context_variables))
                    if "url" in elem_dict:
                        as_url = Source(
                            url=render(elem_dict["url"], context_variables),
                            template=elem_dict["url"],
                            sha256=sha256,
                            md5=md5,
                            context=context_variables,
                        )
                        final_sources.add(as_url)

    return final_sources
