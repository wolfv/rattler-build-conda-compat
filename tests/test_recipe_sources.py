from __future__ import annotations

from pathlib import Path

import pytest
from rattler_build_conda_compat.loader import load_yaml
from rattler_build_conda_compat.recipe_sources import get_all_url_sources, render_all_sources

test_data = Path(__file__).parent / "data"


@pytest.mark.parametrize(
    ("partial_recipe", "expected_output"),
    [
        ("single_source.yaml", ["https://foo.com"]),
        ("multiple_sources.yaml", ["https://foo.com", "https://bar.com"]),
        ("if_then_source.yaml", ["https://foo.com", "https://bar.com"]),
        (
            "outputs_source.yaml",
            ["https://foo.com", "https://bar.com", "https://baz.com", "https://qux.com"],
        ),
    ],
)
def test_recipe_sources(partial_recipe: str, expected_output: list[str]) -> None:
    """Test that the recipe sources are correctly extracted from the recipe"""
    path = Path(f"{Path(__file__).parent}/data/{partial_recipe}")
    recipe = load_yaml(path.read_text())
    assert list(get_all_url_sources(recipe)) == expected_output


def test_multi_source_render(snapshot) -> None:
    jolt_physics = test_data / "jolt-physics" / "sources.yaml"
    variants = (test_data / "jolt-physics" / "ci_support").glob("*.yaml")

    recipe_yaml = load_yaml(jolt_physics.read_text())
    variants = [load_yaml(variant.read_text()) for variant in variants]

    sources = render_all_sources(recipe_yaml, variants)
    assert sources == snapshot


def test_conditional_source_render(snapshot) -> None:
    jolt_physics = test_data / "conditional_sources.yaml"
    # reuse the ci_support variants
    variants = (test_data / "jolt-physics" / "ci_support").glob("*.yaml")

    recipe_yaml = load_yaml(jolt_physics.read_text())
    variants = [load_yaml(variant.read_text()) for variant in variants]

    sources = render_all_sources(recipe_yaml, variants)
    assert len(sources) == 4
    assert sources == snapshot
