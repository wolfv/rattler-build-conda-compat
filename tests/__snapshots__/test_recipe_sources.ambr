# serializer version: 1
# name: test_conditional_source_render
  set({
    Source(url='https://blpapi.bloomberg.com/download/releases/raw/files/blpapi_cpp_3.24.11.1-linux.tar.gz', template='https://blpapi.bloomberg.com/download/releases/raw/files/blpapi_cpp_${{ blpapicpp_linux_version }}-linux.tar.gz', context={'name': 'blpapi', 'version': '3.24.11', 'blpapicpp_linux_version': '3.24.11.1', 'blpapicpp_win_version': '3.24.11.1', 'blpapicpp_osx_version': '3.24.11.1'}, sha256='eb1a9aa834a969a8ccbbb04061274623659a1fb273abda4413b47fe59e7ee412', md5=None),
    Source(url='https://blpapi.bloomberg.com/download/releases/raw/files/blpapi_cpp_3.24.11.1-macos-arm64.tar.gz', template='https://blpapi.bloomberg.com/download/releases/raw/files/blpapi_cpp_${{ blpapicpp_osx_version }}-macos-arm64.tar.gz', context={'name': 'blpapi', 'version': '3.24.11', 'blpapicpp_linux_version': '3.24.11.1', 'blpapicpp_win_version': '3.24.11.1', 'blpapicpp_osx_version': '3.24.11.1'}, sha256='fa96331edf06dd2342cb27771367516296067e94961ec6e600add1c2eed9c41d', md5=None),
    Source(url='https://blpapi.bloomberg.com/download/releases/raw/files/blpapi_cpp_3.24.11.1-windows.zip', template='https://blpapi.bloomberg.com/download/releases/raw/files/blpapi_cpp_${{ blpapicpp_win_version }}-windows.zip', context={'name': 'blpapi', 'version': '3.24.11', 'blpapicpp_linux_version': '3.24.11.1', 'blpapicpp_win_version': '3.24.11.1', 'blpapicpp_osx_version': '3.24.11.1'}, sha256='e63c0b75d50097194c425489a08cec9c0374ab7b8eafe5b015c076bb2432fa19', md5=None),
    Source(url='https://blpapi.bloomberg.com/repository/releases/python/blpapi-3.24.11.tar.gz', template='https://blpapi.bloomberg.com/repository/releases/python/${{ name }}-${{ version }}.tar.gz', context={'name': 'blpapi', 'version': '3.24.11', 'blpapicpp_linux_version': '3.24.11.1', 'blpapicpp_win_version': '3.24.11.1', 'blpapicpp_osx_version': '3.24.11.1'}, sha256='94dc699df262187b3afe4f163526aac67fb5982b008efe851836e9f1cd5358c1', md5=None),
  })
# ---
# name: test_multi_source_render
  set({
    Source(url='https://all.com', template='https://all.com', context={'name': 'foobar', 'version': '1.2.3'}, sha256='yyy', md5=None),
    Source(url='https://all.com/foobar/1.2.3.zip', template='https://all.com/${{ name }}/${{ version }}.zip', context={'name': 'foobar', 'version': '1.2.3'}, sha256='yyy', md5=None),
    Source(url='https://foo.com', template='https://foo.com', context={'name': 'foobar', 'version': '1.2.3'}, sha256='679d42a76b82a6dc20ffd9048ab45fb2e5557a526113aaedc7419e12f3cfab38', md5=None),
    Source(url='https://osx.com/osx-64/foo.zip', template='https://osx.com/${{ target_platform }}/foo.zip', context={'name': 'foobar', 'version': '1.2.3'}, sha256='zzz', md5=None),
    Source(url='https://osx.com/win-64/foo.zip', template='https://osx.com/${{ target_platform }}/foo.zip', context={'name': 'foobar', 'version': '1.2.3'}, sha256='zzz', md5=None),
    Source(url='https://win.com', template='https://win.com', context={'name': 'foobar', 'version': '1.2.3'}, sha256='xxx', md5=None),
    Source(url=['https://foo.com/linux-64/zip.zip', 'https://mirror.com/linux-64/zip.zip'], template=['https://foo.com/${{ target_platform }}/zip.zip', 'https://mirror.com/${{ target_platform }}/zip.zip'], context={'name': 'foobar', 'version': '1.2.3'}, sha256='yyy', md5=None),
    Source(url=['https://foo.com/osx-64/zip.zip', 'https://mirror.com/osx-64/zip.zip'], template=['https://foo.com/${{ target_platform }}/zip.zip', 'https://mirror.com/${{ target_platform }}/zip.zip'], context={'name': 'foobar', 'version': '1.2.3'}, sha256='yyy', md5=None),
    Source(url=['https://foo.com/osx-arm64/zip.zip', 'https://mirror.com/osx-arm64/zip.zip'], template=['https://foo.com/${{ target_platform }}/zip.zip', 'https://mirror.com/${{ target_platform }}/zip.zip'], context={'name': 'foobar', 'version': '1.2.3'}, sha256='yyy', md5=None),
    Source(url=['https://foo.com/win-64/zip.zip', 'https://mirror.com/win-64/zip.zip'], template=['https://foo.com/${{ target_platform }}/zip.zip', 'https://mirror.com/${{ target_platform }}/zip.zip'], context={'name': 'foobar', 'version': '1.2.3'}, sha256='xxx', md5=None),
  })
# ---
